alf/environments/suite_tic_tac_toe.py

```diff
@@ -0,0 +1,192 @@
+             [0, 1, 2], [2, 1, 0]]).unsqueeze(0)
+        self._B = torch.arange(self._batch_size)
+        self._empty_board = self._observation_spec.zeros()
+        self._boards = self._observation_spec.zeros((self._batch_size, ))
+        self._env_ids = torch.arange(batch_size)
+        self._player_0 = torch.tensor(-1.)
+        self._player_1 = torch.tensor(1.)
+
+    @property
+    def batched(self):
+        return True
+
+    @property
+    def batch_size(self):
+        return self._batch_size
+
+    def env_info_spec(self):
+        return {
+            "play0_win": alf.TensorSpec(()),
+            "play1_win": alf.TensorSpec(()),
+            "draw": alf.TensorSpec(()),
+            "invalid_move": alf.TensorSpec(()),
+        }
+
+    def observation_spec(self):
+        return self._observation_spec
+
+    def observation_desc(self):
+        return ""
+
+    def action_spec(self):
+        return self._action_spec
+
+    def _reset(self):
+        self._boards = self._observation_spec.zeros((self._batch_size, ))
+        self._game_over = torch.zeros((self._batch_size, ), dtype=torch.bool)
+        self._prev_action = self._action_spec.zeros((self._batch_size, ))
+        return TimeStep(
+            observation=self._boards.clone().detach(),
+            step_type=torch.full((self._batch_size, ), StepType.FIRST),
+            reward=torch.zeros((self._batch_size, )),
+            discount=torch.ones((self._batch_size, )),
+            prev_action=self._action_spec.zeros((self._batch_size, )),
+            env_id=self._env_ids,
+            env_info={
+                "play0_win": torch.zeros(self._batch_size),
+                "play1_win": torch.zeros(self._batch_size),
+                "draw": torch.zeros(self._batch_size),
+                "invalid_move": torch.zeros(self._batch_size),
+            })
+
+    def _step(self, action):
+        prev_game_over = self._game_over
+        prev_action = action.clone()
+        prev_action[prev_game_over] = 0
+        self._boards[prev_game_over] = self._empty_board
+        step_type = torch.full((self._batch_size, ), int(StepType.MID))
+        player = self._get_current_player().to(torch.float32)
+        x = action % 3
+        y = action // 3
+        valid = self._boards[self._B, y, x] == 0
+        self._boards[self._B[valid], y[valid], x[valid]] = player[valid]
+        won = self._check_player_win(player)
+        reward = torch.where(won, -player, torch.tensor(0.))
+        reward = torch.where(valid, reward, player)
+        game_over = self._check_game_over()
+        game_over = torch.max(game_over, ~valid)
+        step_type[game_over] = int(StepType.LAST)
+        step_type[prev_game_over] = int(StepType.FIRST)
+        discount = torch.ones(self._batch_size)
+        discount[game_over] = 0.
+        self._boards[prev_game_over] = self._empty_board
+        self._game_over = game_over
+        self._prev_action = action
```

Comment:
``self._prev_action = prev_action`` here? Otherwise the rendering might be incorrect.

Reply:
Your are right. There is no issue here.